### DISCUSSION TITLE
Best way to implement "Loading data with signed URLs" in WMS extension

### DISCUSSION THREAD

**Discussion by benjaminleighton** [developmentseed/titiler#700] [Q&A]
I'm wondering if there is an elegant way to implement https://developmentseed.org/titiler/examples/code/working_with_signed_urls/ for the WMS extension? In our fork we modify the WMS extension to expect base64 and then decode that in the layer parameter of the wms function https://github.com/AusClimateService/nbic-titiler/blob/main/src/titiler/extensions/titiler/extensions/wms.py#L288

As suggested in the Titiler documentation, a DatasetPathParams dependency could be implemented to decode a base64 url. However this is insufficient. Clients make subsequent queries based on layers returned in the initial GetCapabilities response, they will be presigned-urls in this use case and therefore need to be base64 encoded. We achieve that by re-encoding the decoded urls back to base64 prior to building the GetCapabilities response https://github.com/AusClimateService/nbic-titiler/blob/main/src/titiler/extensions/titiler/extensions/wms.py#L383. 

I'd like to put in a pull request to enable presigned-url capability for WMS in Titiler but I can't figure out how to rewrite the encoding logic elegantly? Would it be ok to put this in conditional logic so that behaviour is controlled in the initial GetCapabilities request and maybe (if it is possible) in subsequent more specific WMS requests. Perhaps this could be achieved by passing an encoded=True parameter?

**Comment by vincentsarago** [comment]
👋 @benjaminleighton

>  In our fork

You better use titiler python module instead of forking it. This way you can have your own tiler application with you own custom `WMS` extension and no worry about upstream changes.

**Reply by benjaminleighton** [reply]
Thanks @vincentsarago. I'll look into using a module. Would an approach be to put in a pull request that proposes a refactor the WMS extension so it can be imported as a module and then certain logic overriden in a custom WMS extension? Then I could use the module approach reusing bulk of the Titiler WMS extension logic and overriding targeted bits of functionality. The existing Depenencies model provides a good extension point for processing parameters but I don't think it will work for my case of signed urls. 

I could have a go at wholesale duplicating or implementing an alternative WMS extension so that upstream changes aren't a problem but then I'd have to manually adapt any improvements to the existing Titiler WMS extension into our extension. We are keen to contribute back to Titiler itself. A colleague is working on GetLegendGraphic functionality which we think might be useful for both the existing WMS and also, perhaps, as a standaloneTitiler endpoint.
